### ISSUE TITLE
For unbundled builds, be able to configure reference to be inlined.

### ISSUE THREAD

**Issue by justinfagnani** [Polymer/polymer-bundler#533]
_From @bahrus on June 2, 2017 23:37_

### Description
In the scheme of things, this is probably a low priority request (at best).  Just wanted to get it off my chest, in case this makes sense to anyone else.

There is no established protocol for editing / handing typescript  script tags within an html document.  Hence those individuals who want to utilize typescript are pretty much forced to separate the code from the HTML file.  Maybe that's a good thing anyway, and some none typescripters would also prefer to keep things separate during development.

But while debugging in chrome, It would be kind of nice to see the class definition combined in the same "file" as the html template.  That way if you jump to the function definition, you can see the template right there.    Likewise, even for unbundled, HTTP/2 settings, there _might_ be a small performance benefit from inlining the script file into the html file.  The benefit would only be there if the script file is tightly coupled to a single html file, which will typically be the case for Polymer components.

So my proposal is to support at least one of these options:

- Provide a build setting in polymer.json, for unbundled builds, to always inline script files if they are only referenced by a single html file (with the same name?).  Make polymer serve apply the same logic.  Or

- Support some sort of custom attribute on the script tag that indicates the developer wants the script to be embedded in the html during builds (and when serving via polymer serve).  Or

-  Just aggressively do the inlinining without any config setting, for both the unbundled build, and the polymer serve, if there is no question that performance is improved in all cases by doing this, and if the polymer analyzer is 100% confident the script file is tightly coupled to the html file.

_Copied from original issue: Polymer/polymer-cli#783_

**Comment by usergenic** [comment]
@bahrus I'm not familiar enough with the tooling around compiling typescript inside script tags like:
```html
<script type = "text/typescript">
    //TypeScript code goes here
</script>
```

...but it feels like if our html splitter/joiner in `polymer-build` supported embedded TypeScript to enable compilation of ts to js and then inlining it back into the html it came from would be on the order of what you'd want?

**Comment by usergenic** [comment]
We have some outstanding issues with source-maps for babel-compiled code originating from inline script tags esp in bundled output, but assuming we square those away, adding typescript support is something I think we would look at doing.  (We all use TypeScript here on the tools anyways, so first-class TS support for Polymer is a goal for us).

**Comment by bahrus** [comment]
[Webstorm](https://www.jetbrains.com/help/webstorm/2017.1/using-language-injections.html)  does provide tooling support for one language embedded within another.  I've tried it with Typescript, and it works reasonably well.  If Polymer serve and the bundler supported compiling the typescript on the fly (or during the build), that would also be a really nice feature.

The idea isn't [new](https://www.npmjs.com/package/inline-ts), but I think Polymer serve and the bundler are compiling TypeScript / ES6 in a special way in order to support IE11, so that would need to be factored in as well.  Just speculating.

I've switched to VSCode, though, because of its lighter footprint, and a few other reasons.  I'm happily using Polymer's IDE plug-in for it, which is much more robust than WebStorm's current support for Polymer/WebComponents -- another important reason I am happy I made the switch.  The plug-in only works for VSCode and Atom currnently.

Unfortunately, similar support for language injection is something [VS Code has not committed to](https://github.com/Microsoft/vscode/issues/1751).  If they did, that would considerably raise the value of any support Polymer could add to inline Typescript.  

It's unclear to me if the Atom Editor supports language injections.  If either VS Code or Atom supported language injection (or if the Polymer IDE could somehow do this), and if it is (almost) as robust as pure Typescript files, and if Polymer serve and build could handle such  syntax gracefully, that might very well be the preferred approach for TypeScripters like me, at least in some cases.  But that's a lot of if's.

So I think providing some support for inlining separate files would be beneficial short term, certainly, and even longer term for a significant number of people who value the separation of view from compoment model, and/or keeping "risks" separate by segregating schema-backed DSL's from free form code.  I think this principle is quite important to Aurelia fans, for example.  It has some sway with me as well.

Also, in general, I think giving the "unbundled" build a little flexibility in terms of inlining resources, including but not limited to JavaScript (I'm thinking CSS as well for Less/Sass/Stylus support), is something worth considering.
